crd-generator/api/src/main/java/io/fabric8/crd/generator/DeterministicObjectMapper.java

```diff
@@ -0,0 +1,206 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.crd.generator;
+
+import com.fasterxml.jackson.databind.BeanProperty;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.MapperFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.fasterxml.jackson.databind.ser.std.StdDelegatingSerializer;
+import com.fasterxml.jackson.databind.util.Converter;
+import com.fasterxml.jackson.databind.util.StdConverter;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Method;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_EMPTY;
+import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL;
+import static com.fasterxml.jackson.annotation.JsonInclude.Value.construct;
+
+/**
+ * A {@link ObjectMapper} that serializes YAML in a deterministic way.
+ *
+ * Converts unsorted collections into sorted lists, using the provided converters.
+ */
+class DeterministicObjectMapper {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DeterministicObjectMapper.class);
+
+  static DeterministicObjectMapperBuilder builder() {
+    return new DeterministicObjectMapperBuilder();
+  }
+
+  static class DeterministicObjectMapperBuilder {
+
+    private final CustomComparators customComparators = new CustomComparators();
+
+    private DeterministicObjectMapperBuilder() {
+    }
+
+    <T, U extends Comparable<? super U>> DeterministicObjectMapperBuilder withConverter(final Class<T> clazz,
+        final Function<? super T, ? extends U> keyExtractor) {
+      return withConverter(clazz, Comparator.comparing(keyExtractor));
+    }
+
+    <T> DeterministicObjectMapperBuilder withConverter(final Class<T> clazz, final Comparator<T> comparator) {
+      customComparators.addConverter(clazz, comparator);
+      return this;
+    }
+
+    ObjectMapper build() {
+      final ObjectMapper mapper = JsonMapper.builder(new YAMLFactory()
```

Comment:
This mapper is configured slightly differently from before, more specifically:

`YAML_MAPPER.configure(SerializationFeature.WRITE_NULL_MAP_VALUES, false)` and `YAML_MAPPER.configure(SerializationFeature.WRITE_EMPTY_JSON_ARRAYS, false);` are missing.

Reply:
Both are deprecated API, so I looked how to configure it with up to date API.

`YAML_MAPPER.configure(SerializationFeature.WRITE_NULL_MAP_VALUES, false)` is `.withConfigOverride(Map.class, configOverride -> configOverride.setInclude(construct(NON_NULL, NON_NULL)))`.

`YAML_MAPPER.configure(SerializationFeature.WRITE_EMPTY_JSON_ARRAYS, false);` is `.serializationInclusion(NON_EMPTY)`.

Reply:
Nice 👍 , thanks for the explanation!